pkg/agent/controller/serviceimport.go

```diff
@@ -0,0 +1,224 @@
+			}
+		}()
+	}
+}
+
+func (c *ServiceImportController) serviceImportCreatedOrUpdated(obj interface{}, key string) {
+	if _, found := c.endpointControllerCreated.Load(key); found {
+		klog.V(2).Infof("The endpointController is already running for  key %q", key)
+		return
+	}
+
+	serviceImportCreated := obj.(*lighthousev2a1.ServiceImport)
+	if serviceImportCreated.Spec.Type != lighthousev2a1.Headless {
+		return
+	}
+
+	annotations := serviceImportCreated.ObjectMeta.Annotations
+	serviceNameSpace := annotations[originNamespace]
+	serviceName := annotations[originName]
+	var service *corev1.Service
+	var isFound bool
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var err error
+		service, err = c.kubeClientSet.CoreV1().Services(serviceNameSpace).Get(serviceName, metav1.GetOptions{})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				isFound = false
+			}
+			return err
+		}
+		isFound = true
+		return nil
+	})
+
+	if retryErr != nil || !isFound {
+		klog.Errorf("Error retrieving the service  %q from the namespace %q : %v", serviceName, serviceNameSpace, retryErr)
+		return
+	}
+
+	if service.Spec.Selector == nil {
+		klog.Errorf("The service %q in the namespace %q without selector is not supported", serviceName, serviceNameSpace)
+		return
+	}
+
+	labelSelector := labels.Set(service.Spec.Selector).AsSelector()
+	endpointController, err := NewEndpointController(c.kubeClientSet, serviceImportCreated.ObjectMeta.UID,
+		serviceImportCreated.ObjectMeta.Name, c.clusterID)
+
+	if err != nil {
+		klog.Errorf("Creating endpoint controller for service %q in the namespace %q failed", serviceName, serviceNameSpace)
+		return
+	}
+
+	err = endpointController.Start(endpointController.stopCh, labelSelector)
+	if err != nil {
+		klog.Errorf("Starting endpoint controller for service %q in the namespace %q failed", serviceName, serviceNameSpace)
+		return
+	}
+
+	c.endpointControllerCreated.Store(key, endpointController)
+}
+
+func (c *ServiceImportController) serviceImportDeleted(key string) {
+	obj, found := c.endpointControllerDeleted.Load(key)
+	if !found {
+		klog.Errorf("deleting endpointSlice for key %s failed since object not found", key)
+		return
+	}
+
+	c.endpointControllerDeleted.Delete(key)
+
+	si := obj.(lighthousev2a1.ServiceImport)
+	matchLabels := si.ObjectMeta.Labels
+	labelSelector := labels.Set(map[string]string{"app": matchLabels["app"]}).AsSelector()
+	err := c.kubeClientSet.DiscoveryV1beta1().EndpointSlices(si.Namespace).
+		DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: labelSelector.String()})
+	if err != nil && !errors.IsNotFound(err) {
+		c.endpointControllerDeleted.Store(key, si)
+		c.queue.AddRateLimited(key)
+
+		return
+	}
+
+	if obj, found := c.endpointControllerCreated.Load(key); found {
+		endpointController := obj.(*EndpointController)
+		endpointController.endPointqueue.ShutDown()
+		close(endpointController.stopCh)
+		c.endpointControllerCreated.Delete(key)
+	}
```

Comment:
I think we should stop the controller regardless of whether or not the delete above fails, ie move this before the delete.

Reply:
Done.